### ISSUE TITLE
Drupal Version Exposed in Build Artifact

### ISSUE THREAD

**Issue by aweingarten** [acquia/blt#3476]
When I scan a stock blt'd site running BLT 10 with:
https://whatcms.org

I get back a drupal version including minor version.  We are running 8.6.10.  This returns 8.6.9. I wonder what its able to use to fingerprint the version but its a little too close for comfort.

**Comment by mikemadison13** [comment]
FYI I just checked a site (https://whatcms.org/?s=www.drupalgovcon.org) that is definitely running Drupal 8.6.10 and that site is showing it as 8.6.9... so i'm skeptical of what methods it is using to detect things at all. Govcon is still running BLT 9.2.x though, so maybe there is something in 10 that is different?

**Comment by aweingarten** [comment]
So I ssh'd into our Acquia server and ran the following:

`grep -rnw '.' -e '8.6.10'  --exclude=*.svg --exclude-dir=themes`

saw the following
```./core/lib/Drupal.php:85:  const VERSION = '8.6.10';
./modules/contrib/lightning_core/CHANGELOG.md:2:* Security updated Drupal core to 8.6.10 (SA-CORE-2019-003).
./modules/contrib/lightning_core/composer.json:11:        "drupal/core": "~8.6.10",
```

**Comment by danepowell** [comment]
CHANGELOG.md gets stripped in BLT 10.0.0-beta3 and 9.2.4. You must not be running the latest version. I cannot reproduce this on fresh builds of either release.

Please reopen if you can provide testing steps that reproduce this using fresh builds of 10.0.0-beta3 or 9.2.4.

**Comment by aweingarten** [comment]
@danepowell, so I checked its not coming from that lightning change log.  Can you please reopen.

**Comment by aweingarten** [comment]
Looking here to see what could have changed between 8.6.8 and 8.6.9 that could have given a fingerprint:

https://github.com/drupal/drupal/compare/8.6.8..8.6.9.patch

**Comment by aweingarten** [comment]
So the fact that govcon is showing 8.6.9 instead of 8.6.10 suggests to me that this specific version isn't in the code. I'm guessing that the fingerprinter is picking up one of the following from the 8.6.8 to the 8.6.9 patch [here]( https://github.com/drupal/drupal/compare/8.6.8..8.6.9.patch): 
- https://www.drupalgovcon.org/core/misc/ajax.es6.js
- https://www.drupalgovcon.org/core/misc/ajax.js
- https://www.drupalgovcon.org/core/modules/system/tests/fixtures/update/drupal-8.6.0-minimal-with-warm-caches.sql.gz

Any of these files could fingerprint 8.6.9.  

There is also some other identifying files like:
- https://www.drupalgovcon.org/core/yarn.lock
- https://www.drupalgovcon.org/core/phpcs.xml.dist

The breathe of this issue _may_ be out of the scope of BLT.   

However, it looks like there may be a false sense of security in the current blt scheme to sanitize files with drupal versions from the Drupal Codebase.

cc @CashWilliams

**Comment by aweingarten** [comment]
Given this some more thought.

I wonder if there is a simple htaccess rule that can be written to block access to static files in the /core directory.  That should block all core fingerprinting. 

The only static files I can think of that could be a problem would be .js. These could be initially mitigated by turning on aggregation. Maybe some images.

Longer term there may need to be a php obfuscatory that operates on core js.

**Comment by danepowell** [comment]
It sounds like a nice-to-have, but it would need to be done in a way that minimizes potential impact for less paranoid users. For instance, many people in the past have proposed stripping _all_ `.txt` and `.md` files from the deploy artifact, but it's easy to show how that would break certain modules (#3423).

BLT does not claim to make your deploy artifact impossible to fingerprint. The sanitization has always been on a best-effort basis.

**Comment by dpagini** [comment]
Could maybe the deploy santize command use a fileset? We can provide a default fileset in BLT and `config.yml`, and that could be over-ridden per project by providing a different fileset?

**Comment by danepowell** [comment]
I don't think we have any mechanism to define a fileset via configuration right now. That sounds like a heavy lift for a small payoff. PRs welcome, but again this would need to not interfere with existing projects.

I would recommend anyone needing this functionality right now simply implement a post-deploy-build hook, or modify the deploy .gitignore file to exclude the files you don't want.

**Comment by dpagini** [comment]
> I don't think we have any mechanism to define a fileset via configuration right now. 

Not sure if we're talking about something different here, but aren't basically all the validate functions taking in a fileset object?
https://github.com/acquia/blt/blob/10.x/config/build.yml#L241-L258

**Comment by danepowell** [comment]
Yeah but those are just the names of hardcoded fileset definitions. You can't modify what files are included as part of those filesets.

For instance, here's where we define what constitutes PHP files: https://github.com/acquia/blt/blob/c89b95ccea6426d39efee1e0b8c8d3b0b77386be/src/Robo/Filesets/Filesets.php#L102

**Comment by dpagini** [comment]
Right... so there could be a config entry for...
```
artifact:
  scrub:
    filesets:
      - files.artifact.scrub
```

Then in `blt/src/Robo/Filesets/Filesets.php` we define `files.artifact.scrub` to be the same finder used in the `DeployCommand` today. Then, as a user who wants to extend that, I could, for example, change my project config to be...

`blt/blt.yml`
```
artifact:
  scrub:
    filesets:
      - files.artifact.scrub.custom
```

...or something like that. That honestly may not be much more or less work than just overridding the Deploy command. Was just trying to throw in a suggestion. I've done the following before to have additional twig files validated.

```
validate:
  twig:
    filesets: [files.twig, files.twig.custom]
```

Or to go FURTHER... BLT could define fileset for something like "all .txt files" and let users choose to enable it. So for example, the default config would look like:

```yml
artifact:
  scrub:
    filesets:
      - files.artifact.scrub.drupal-core-static
      - files.artifact.scrub.vcs-dirs
      # Copy this block to your project BLT config and uncomment the below to sanitize these files.
      # - files.artifact.scrub.all-txt-files
```

**Comment by aweingarten** [comment]
BTW for the record I consulted with the D.O security team they do not consider this a security issue which is a little frustrating.

**Comment by aweingarten** [comment]
TLDR.  I honestly don't think BLT is the place to solve for this. @danepowell, @mikemadison13 despite what everyone thinks I don't believe that everything is your responsibility

I wonder if this should be a new contrib module + htaccess patch checker.  @dpagini side project ?

**Comment by greylabel** [comment]
I see their point, removing these core files does not make the application code more secure.

**Comment by alexxed** [comment]
I don't see how this could be completely solved in BLT since we don't know what the scanners are looking for. They may look for files with the version, or they could be smart enough to follow the drupal change log and find there a way to tell the difference between versions. A customer may also have some modules that expose that.

I did a CMS scanner years ago, I was only interested in the CMS type, not version, so I know people can get very ingenious with the scanners.

You can play around with a Drupal installation and reverse engineer whatcms by removing stuff until it fails to detect. But that's one tool; there are others.

**Comment by danepowell** [comment]
I think we can all agree that while this might be desirable for some customers, it's not within the scope of BLT. Sounds like a good use case for a new package that you could integrate with BLT via the post-deploy-build hook. If there's something else that BLT could do to make that integration easier, feel free to open a new issue.
